styx-common/src/main/java/com/spotify/styx/util/ShardedCounter.java

```diff
@@ -250,20 +338,32 @@ public static long getLimit(Datastore datastore, String counterId) {
    */
   public void updateCounter(DatastoreReaderWriter transaction, String counterId, long delta) {
     CounterSnapshot snapshot = getCounterSnapshot(counterId);
+    if (getCounter(counterId) >= snapshot.limit) {
+      // when counter usage is at/over its limit, only allow negative deltas
+      if (delta < 0) {
+        int shardIndex = snapshot.pickShardWithExcessUsage(delta);
+        updateCounterShard(transaction, snapshot, shardIndex, delta);
+        return;
+      }
+      throw new CounterCapacityException("Counter %s has no more capacity.", counterId);
+    }
+
     int shardIndex = snapshot.pickShardWithSpareCapacity(delta);
+    updateCounterShard(transaction, snapshot, shardIndex, delta);
```

Comment:
Could refactor this method, these lines are a repeat.

Reply:
If you're referring to L351-352: only L352 repeats on L345, but the shardIndex is determined by two different methods, and I find this repetition to be more readable.